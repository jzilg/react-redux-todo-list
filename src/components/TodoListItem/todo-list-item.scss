@import "../../style/color";

.todo {
    display: grid;
    grid-template-columns: 1.6fr 1fr 2fr .6fr .8fr;
    grid-column-gap: 10px;
    align-items: center;
    height: 100%;
}

.input {
    height: 100%;
    border: 0;
    background-color: transparent;
    text-overflow: ellipsis;
}

.last-event-input {
    margin-left: 4px;
    @extend .input;
}

.schedule-input {
    width: 42px;
    margin-right: 5px;
    margin-left: 5px;
    text-align: center;
    @extend .input;
}

.btn-list {
    display: flex;
    height: 100%;
    margin-top: 0;
    margin-bottom: 0;
    padding-left: 0;
    list-style: none;
}

.btn-list-item {
    flex-grow: 1;
}

.save-btn,
.delete-btn {
    width: 100%;
    height: 100%;
    padding: 0;
    border-width: 0;
    background-color: transparent;
}

.save-btn {
    transition: background-color .2s, color .2s, opacity .2s;

    &[disabled] {
        opacity: 0;
    }

    &:hover {
        background-color: $green;
        color: #fff;
    }
}

.delete-btn {
    font-size: 24px;
    color: $grey;
    transition: background-color .2s, color .2s;

    &:hover {
        background-color: $red;
        color: #fff;
    }
}
