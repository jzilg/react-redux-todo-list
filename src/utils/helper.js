import moment from 'moment'

/**
 * @returns {string} todays date (YYYY-MM-DD)
 */
export function getTodaysDate() {
    return moment().format('YYYY-MM-DD')
}

/**
 * @param {string} startDate moment date format
 * @param {string} endDate moment date format
 * @returns {number} duration in days
 */
export function getDurationBetweenDates(startDate, endDate) {
    const startDateMoment = moment(startDate)
    const endDateMoment = moment(endDate)
    return moment.duration(startDateMoment.diff(endDateMoment)).asDays()
}
