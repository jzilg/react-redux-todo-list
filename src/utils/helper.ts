import moment from 'moment'

/**
 * @returns {string} todays date (YYYY-MM-DD)
 */
export function getTodaysDate(): string {
    return moment().format('YYYY-MM-DD')
}

/**
 * @param {string} startDate moment date format
 * @param {string} endDate moment date format
 * @returns {number} duration in days
 */
export function getDurationBetweenDates(startDate, endDate): number {
    const startDateMoment = moment(startDate)
    const endDateMoment = moment(endDate)
    return startDateMoment.diff(endDateMoment, 'days')
}

function randomIntFromInterval(min: number, max: number): number {
    const { random, floor } = Math
    return floor(random() * (max - min + 1) + min)
}

/**
 * @returns {number} new id base64 of timestamp and random number
 */
export function createUniqueId(): number {
    const date = new Date()
    const time = date.getTime()
    const randomNumber = randomIntFromInterval(10000, 99999)
    return Math.floor(time / randomNumber)
}
