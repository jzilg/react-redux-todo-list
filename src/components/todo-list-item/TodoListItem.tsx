import React, { ReactNode } from 'react'
import equal from 'deep-equal'
import Todo from '../../entities/todo.interface'
import Icon from '../icon'
import Urgency from '../urgency'
import { getTodaysDate } from '../../utils/helper'
import style from './todo-list-item.scss'

interface Props {
    todo: Todo
    saveTodo: Function
    removeTodo: Function
    isLoading: boolean
}

interface State extends Todo {}

class TodoListItem extends React.Component<Props, State> {
    today = getTodaysDate()

    constructor(props) {
        super(props)

        const { todo } = props

        this.state = {
            id: todo.id,
            name: todo.name,
            schedule: parseInt(todo.schedule, 10),
            lastEvent: todo.lastEvent || this.today,
        }

        this.inputChange = this.inputChange.bind(this)
        this.saveTodo = this.saveTodo.bind(this)
        this.removeTodo = this.removeTodo.bind(this)
        this.setInputToday = this.setInputToday.bind(this)
    }

    setInputToday(): void {
        const { state } = this
        if (state.lastEvent === this.today) {
            return
        }

        this.setState({
            lastEvent: this.today,
        }, () => {
            this.saveTodo()
        })
    }

    /**
     * @returns {boolean}
     */
    todoHasChanged(): boolean {
        const { todo } = this.props
        return !equal(todo, this.state)
    }

    /**
     * @param {Object} event
     */
    inputChange(event): void {
        const { name } = event.target
        let { value } = event.target

        if (name === 'schedule') {
            value = parseInt(value, 10)
        }

        this.setState(state => ({
            ...state,
            [name]: value,
        }))
    }

    saveTodo(): void {
        const { saveTodo } = this.props
        saveTodo(this.state)
    }

    removeTodo(): void {
        const { removeTodo } = this.props
        removeTodo(this.state)
    }

    render(): ReactNode {
        const { isLoading } = this.props
        const { name, schedule, lastEvent } = this.state
        const saveBtnIsDisabled = (): boolean => !this.todoHasChanged() || isLoading
        const saveBtnTitle = saveBtnIsDisabled() ? '' : 'Save TodoListItem'

        return (
            <form className={style.todo} title={name}>
                <input
                    type="text"
                    name="name"
                    value={name}
                    placeholder="Name"
                    autoComplete="off"
                    onChange={this.inputChange}
                    autoFocus={!name.length}
                    className={style.input}
                />
                <div>
                    <span>
                        Every
                    </span>
                    <input
                        type="number"
                        name="schedule"
                        className={style['schedule-input']}
                        value={schedule}
                        min="1"
                        autoComplete="off"
                        onChange={this.inputChange}
                    />
                    <span>
                        {schedule === 1 ? 'Day' : 'Days'}
                    </span>
                </div>
                <div>
                    <span>
                        Last done on
                    </span>
                    <input
                        type="date"
                        name="lastEvent"
                        className={style['last-event-input']}
                        value={lastEvent}
                        onChange={this.inputChange}
                    />
                    <button
                        type="button"
                        onClick={this.setInputToday}
                    >
                        Today
                    </button>
                </div>
                <Urgency
                    today={this.today}
                    lastEvent={lastEvent}
                    schedule={schedule}
                />
                <ul className={style['btn-list']}>
                    <li className={style['btn-list-item']}>
                        <button
                            type="button"
                            className={style['save-btn']}
                            onClick={this.saveTodo}
                            disabled={saveBtnIsDisabled()}
                            title={saveBtnTitle}
                        >
                            <Icon name="save" width="1em" height="auto" />
                        </button>
                    </li>
                    <li className={style['btn-list-item']}>
                        <button
                            type="button"
                            className={style['delete-btn']}
                            onClick={this.removeTodo}
                            disabled={isLoading}
                            title="Remove Todo"
                        >
                            âœ•
                        </button>
                    </li>
                </ul>
            </form>
        )
    }
}

export default TodoListItem
