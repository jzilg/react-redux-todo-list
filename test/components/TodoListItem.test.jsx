import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import TodoListItem from '../../src/components/todo-list-item'
import { getTodaysDate } from '../../src/utils/date'

describe('TodoListItem', () => {
    const today = '2018-12-04'
    const todo = {
        id: 1,
        name: 'Papa anrufen',
        schedule: 4,
        lastEvent: '2018-06-03',
    }

    it('should render correctly one todo if duration between lastEvent and schedule is > 1', () => {
        // pretend today is the date when the snapshot was taken
        Date.now = () => 1543881600000 // 2018-12-04

        const removeTodo = jest.fn()
        const saveTodo = jest.fn()

        const component = shallow((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })

    it('should render correctly one todo if duration between lastEvent and schedule is == 1', () => {
        // pretend today is the date when the snapshot was taken
        Date.now = () => 1543881600000 // 2018-12-04

        const removeTodo = jest.fn()
        const saveTodo = jest.fn()
        const todoWithLastEventYesterday = {
            ...todo,
            lastEvent: '2018-12-03',
            schedule: 1,
        }

        const component = shallow((
            <TodoListItem
                todo={todoWithLastEventYesterday}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })

    it('should call saveTodo if save-button is clicked', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()

        const component = shallow((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        component.find('button').at(1).simulate('click')
        expect(saveTodo.mock.calls.length).toBe(1)
    })

    it('should call removeTodo if remove-button is cliced', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()

        const component = shallow((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        component.find('button').last().simulate('click')
        expect(removeTodo.mock.calls.length).toBe(1)
    })

    it('should not call saveTodo if setInputToday is called but last event is today', () => {
        // pretend today is the date when the snapshot was taken
        Date.now = () => 1543881600000 // 2018-12-04

        const removeTodo = jest.fn()
        const saveTodo = jest.fn()

        const todoWithLastEventToday = {
            ...todo,
            lastEvent: today,
        }
        const component = shallow((
            <TodoListItem
                todo={todoWithLastEventToday}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        component.find('button').first().simulate('click')
        expect(saveTodo.mock.calls.length).toBe(0)
    })

    it('should set lastEvent if lastEvent-input changes', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()
        const newLastEvent = '2018-03-20'
        const event = {
            target: {
                value: newLastEvent,
            },
        }

        const component = mount((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        component.find('input').at(2).simulate('change', event)
        component.update()

        const name = component.find('input').at(2).props().value
        expect(name).toBe(newLastEvent)
    })

    it('should set lastEvent to today, if setInputToday is called and last event was in the past', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()

        const component = mount((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        component.find('button').first().simulate('click')
        const lastEvent = component.find('input').at(2).props().value
        expect(lastEvent).toBe(today)
        expect(saveTodo.mock.calls.length).toBe(1)
    })

    it('should set name if name-input changes', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()
        const newName = 'Foo'
        const event = {
            target: {
                value: newName,
            },
        }

        const component = mount((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        component.find('input').first().simulate('change', event)
        component.update()

        const name = component.find('input').first().props().value
        expect(name).toBe(newName)
    })

    it('should set schedule if schedule-input changes', () => {
        const removeTodo = jest.fn()
        const saveTodo = jest.fn()
        const newSchedule = 3
        const event = {
            target: {
                value: newSchedule,
            },
        }

        const component = mount((
            <TodoListItem
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        component.find('input').at(1).simulate('change', event)
        component.update()

        const name = component.find('input').at(1).props().value
        expect(name).toBe(newSchedule)
    })
})
