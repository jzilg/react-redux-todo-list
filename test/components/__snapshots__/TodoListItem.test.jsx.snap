// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should render correctly one todo if duration between lastEvent and schedule is == 1 1`] = `
<form
  title="Papa anrufen"
>
  <input
    autoComplete="off"
    autoFocus={false}
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value="Papa anrufen"
  />
  <div>
    <span>
      Every
    </span>
    <input
      autoComplete="off"
      min="1"
      name="schedule"
      onChange={[Function]}
      type="number"
      value={1}
    />
    <span>
      Day
    </span>
  </div>
  <div>
    <span>
      Last done on
    </span>
    <input
      name="lastEvent"
      onChange={[Function]}
      type="date"
      value="2018-12-03"
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Today
    </button>
  </div>
  <Urgency
    lastEvent="2018-12-03"
    schedule={1}
    today="2018-12-04"
  />
  <ul>
    <li>
      <button
        disabled={true}
        onClick={[Function]}
        title=""
        type="button"
      >
        <Icon
          className=""
          height="auto"
          name="save"
          width="1em"
        />
      </button>
    </li>
    <li>
      <button
        disabled={false}
        onClick={[Function]}
        title="Remove Todo"
        type="button"
      >
        ✕
      </button>
    </li>
  </ul>
</form>
`;

exports[`TodoListItem should render correctly one todo if duration between lastEvent and schedule is > 1 1`] = `
<form
  title="Papa anrufen"
>
  <input
    autoComplete="off"
    autoFocus={false}
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value="Papa anrufen"
  />
  <div>
    <span>
      Every
    </span>
    <input
      autoComplete="off"
      min="1"
      name="schedule"
      onChange={[Function]}
      type="number"
      value={4}
    />
    <span>
      Days
    </span>
  </div>
  <div>
    <span>
      Last done on
    </span>
    <input
      name="lastEvent"
      onChange={[Function]}
      type="date"
      value="2018-06-03"
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Today
    </button>
  </div>
  <Urgency
    lastEvent="2018-06-03"
    schedule={4}
    today="2018-12-04"
  />
  <ul>
    <li>
      <button
        disabled={true}
        onClick={[Function]}
        title=""
        type="button"
      >
        <Icon
          className=""
          height="auto"
          name="save"
          width="1em"
        />
      </button>
    </li>
    <li>
      <button
        disabled={false}
        onClick={[Function]}
        title="Remove Todo"
        type="button"
      >
        ✕
      </button>
    </li>
  </ul>
</form>
`;
